hips/hip-9999.md

```diff
@@ -0,0 +1,76 @@
+---
+hip: 9999
+title: "Document backwards-compatibility rules"
+authors: [ "Marc Khouzam <marc.khouzam@montreal.ca>" ]
+created: "2020-09-18"
+type: "process"
+status: "draft"
+---
+
+## Abstract
+
+Define and document the backwards-compatibility rules that apply to Helm minor and patch releases.
+
+## Motivation
+
+Helm aims to strictly respect [Semantic Versioning][semver], where minor and patch releases must be 100% backwards-compatible.
+Backwards-compatibility implies that when moving to a new backwards-compatible release:
+* tools (such as Continuous Integration tools) using the helm CLI will not require any changes
+* programs using the Helm Go API will continue to compile 
+
+Some contributors to the Helm code base may not be comfortable with the details of what constitutes a compatibility-breaking change.
+Furthermore, in some cases, there can be a difference of opinion, even amongst the maintainers, on what constitutes such a change.
+
+Having a well-defined, documented set of backwards-compatibility rules for the Helm code-base will allow to:
+* avoid compatibility-breaking changes being merged by mistake
+* guide contributors in understanding helm's backwards-compatibility requirements
+* remove any uncertainty when it comes to enforcing backwards compatibility
+* provide a basis for discussion and possible updates to how helm handles backwards compatibility
+
+## Rationale
+
+The maintainers have agreed that a formal definition of Helm's backwards-compatibility rules would be beneficial.
+
+## Specification
+
+The following compatibility rules will apply to helm minor and patch releases:
+* exported Go APIs will remain compatible as per this [article][go-module-comp]
+* CLI commands:
+    * are case-sensitive
+    * must not be removed
+    * must not be renamed
+    * must not be moved elsewhere in the command hierarchy
```

Comment:
- Grouping commands (e.g. `helm repo` or `helm plugin` must display only help text if executed as commands. They must not perform any other operations. (Reason to reject PRs)
- Subcommands cannot be "repurposed" to provide a different behavior than the original (as was attempted once when one subcommand was deprecated)

Reply:
> Grouping commands (e.g. helm repo or helm plugin must display only help text if executed as commands. They must not perform any other operations. (Reason to reject PRs)

I agree with this point, but I feel it is not technically a backwards-compatibility issue but more of a helm design rule.  Not sure where it should be documented. Do we (eventually) need a broader document that will describe design rules (or maybe there is one already?) and which may include the backwards-compatibility section? 

>Subcommands cannot be "repurposed" to provide a different behavior than the original (as was attempted once when one subcommand was deprecated)

Good point. I've applied it to commands in general (even though more likely to be attempted on subcommands).

Reply:
In fact, I've applied the "no repurpose" rule to flags as well (and merged both sections)